src/lib/structures/profiling.ml

```diff
@@ -45,58 +45,79 @@ type inst_info = {
 }
 
 type t = {
-  decisions : int ref;
-  assumes : int ref;
-  assumes_current_lvl : int ref;
-  queries : int ref;
-  instantiation_rounds : int ref;
-  instances : int ref;
-  decision_lvl : int ref;
-  instantiation_lvl : int ref;
-
+  mutable decisions : int;
+  mutable assumes : int;
+  mutable assumes_current_lvl : int;
+  mutable queries : int;
+  mutable instantiation_rounds : int;
+  mutable instances : int;
+  mutable decision_lvl : int;
+  mutable instantiation_lvl : int;
   (* 4 kinds of conflicts *)
-  th_conflicts : int ref;
-  b_conflicts : int ref;
-  bcp_th_conflicts : int ref;
-  bcp_b_conflicts : int ref;
-  bcp_mix_conflicts : int ref;
-
+  mutable th_conflicts : int;
+  mutable b_conflicts : int;
+  mutable bcp_th_conflicts : int;
+  mutable bcp_b_conflicts : int;
+  mutable bcp_mix_conflicts : int;
   (* 4 kinds of red/elim *)
-  t_red : int ref;
-  b_red : int ref;
-  t_elim : int ref;
-  b_elim : int ref;
-  (* first int: counter ok kept instances,
-     second int: counter of removed instances*)
-  instances_map : inst_info MS.t ref;
-  instances_map_printed : bool ref
+  mutable t_red : int;
+  mutable b_red : int;
+  mutable t_elim : int;
+  mutable b_elim : int;
+  mutable instances_map : inst_info MS.t;
+  mutable instances_map_printed : bool;
+  timers : Timers.t
 }
 
-let state = {
-  decisions = ref 0;
-  assumes = ref 0;
-  assumes_current_lvl = ref 0;
-  queries = ref 0;
-  instantiation_rounds = ref 0;
-  instances = ref 0;
-  decision_lvl = ref 0;
-  instantiation_lvl = ref 0;
-
-  th_conflicts = ref 0;
-  b_conflicts = ref 0;
-  bcp_th_conflicts = ref 0;
-  bcp_b_conflicts = ref 0;
-  bcp_mix_conflicts = ref 0;
-
-  t_red  = ref 0;
-  b_red  = ref 0;
-  t_elim = ref 0;
-  b_elim = ref 0;
-
-  instances_map = ref MS.empty;
-  instances_map_printed = ref false
+type mode =
+  | Stats
+  | Timers
+  | CallTree
+  | FunctionsTimers
+  | Instances
+
+let mode = ref Stats
+
+let max_nb_prints = 30
+
+let nb_prints = ref max_nb_prints
+
+let initial_info = ref true
+
+let state = lazy {
+  decisions = 0;
+  assumes = 0;
+  assumes_current_lvl = 0;
+  queries = 0;
+  instantiation_rounds = 0;
+  instances = 0;
+  decision_lvl = 0;
+  instantiation_lvl = 0;
+
+  th_conflicts = 0;
+  b_conflicts = 0;
+  bcp_th_conflicts = 0;
+  bcp_b_conflicts = 0;
+  bcp_mix_conflicts = 0;
+
+  t_red = 0;
+  b_red = 0;
+  t_elim = 0;
+  b_elim = 0;
+
+  instances_map = MS.empty;
+  instances_map_printed = false;
+  timers = Timers.empty ()
 }
 
+(** Calls the continuation only when the profiling is on.
+    This allows to only force the state when the profiling is on. *)
+let (let*) state f =
+  if Options.get_profiling () then
+    f (Lazy.force state)
```

Comment:
I did not notice the use of `let*` earlier, apologies — I think that using `let*` here is quite confusing because it leads to `let* state = state in` whose behavior is unclear. `let*` is also traditionally used for a monadic `bind` operator, which this is not.

Why not instead call this something like `if_profiling` or `when_profiling`:

```ocaml
let if_profiling f = if Options.get_profiling () then f (Lazy.force state)
```

and then use it with:

```ocaml
if_profiling @@ fun state ->
...
```

which makes the intent more precise?

Reply:
(As an aside: in the current state of the code, this leads to duplicate `Options.get_profiling ()` tests because most calls to the `Profiling` module are guarded with `if Options.get_profiling`, which is required in some cases because computation has to be performed prior to calling the corresponding profiling function)

Reply:
I usually go for custom `let`s to avoid the extra indentation. It really is a taste choice, I don't mind going for explicit continuations

Reply:
You don't have to use extra indentation with `@@ fun _ ->`, and most styles I am aware of don't (in fact, ocp-indent will not let you use extra indentation in this repo if you use this construct).

Reply:
Oh my bad, I was convinced the `@@` to add indentation! Okay, I'll go for it then